core/src/main/scala/kafka/controller/ControllerChannelManager.scala

```diff
@@ -72,30 +77,37 @@ class ControllerChannelManager (private val controllerContext: ControllerContext
 
   def removeBroker(brokerId: Int) {
     brokerLock synchronized {
-      removeExistingBroker(brokerId)
+      removeExistingBroker(brokerStateInfo(brokerId))
     }
   }
 
   private def addNewBroker(broker: Broker) {
     val messageQueue = new LinkedBlockingQueue[(RequestOrResponse, (RequestOrResponse) => Unit)]()
-    debug("Controller %d trying to connect to broker %d".format(config.brokerId,broker.id))
-    val brokerEndPoint = broker.getBrokerEndPoint(config.interBrokerSecurityProtocol)
-    val channel = new BlockingChannel(brokerEndPoint.host, brokerEndPoint.port,
-      BlockingChannel.UseDefaultBufferSize,
-      BlockingChannel.UseDefaultBufferSize,
-      config.controllerSocketTimeoutMs)
-    val requestThread = new RequestSendThread(config.brokerId, controllerContext, broker, messageQueue, channel)
+    debug("Controller %d trying to connect to broker %d".format(config.brokerId, broker.id))
+    val selector = new Selector(
+      config.socketRequestMaxBytes,
+      config.connectionsMaxIdleMs,
+      metrics,
+      time,
+      //FIXME Replace random suffix by something better
+      s"controller-${config.brokerId}-broker-${broker.id}-${scala.util.Random.nextInt(1000)}",
```

Comment:
Lets see if I got it.

We have a communication thread per broker, and a selector per thread.
If a broker dies, we close the thread and the selector.
If the broker comes back, we can't create a new selector with the same name due to metrics.

Since the number of possible brokers is limited, can we just store the selector for each broker in a Map and reuse / reconnect if the dead broker is revived?